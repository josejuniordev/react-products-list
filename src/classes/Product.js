export default class Product {
  constructor(
    data
  ) {
    this.shortDescription = data.decricaoCurta;
    this.longDescription = data.descricaoLonga;
    this.exclusive = data.exclusivo;
    this.datasheet = data.fichaTecnica;
    this.id = data.id || +new Date();
    this.imageUrl = data.imagem;
    this.name = data.nome;
    this.onSale = data.promocao;
    this.amount = data.valor;
  }

}
