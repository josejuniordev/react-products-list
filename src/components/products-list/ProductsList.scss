@import "../../styles/bem";

.products-list {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;

  &:after {
    content: '';
    width: 30%;
    height: 1px !important;
    justify-self: flex-start;
  }

  @include element(item) {
    width: 30%;
    max-width: 350px;
    margin-bottom: 30px;
    display: flex;
    justify-content: stretch;
  }
}

