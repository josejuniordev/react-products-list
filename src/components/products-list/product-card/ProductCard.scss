@import "../../../styles/colors";
@import "../../../styles/bem";

.product-card {
  display: flex;
  flex-direction: column;

  @include element(image-container) {
    background: center no-repeat;
    background-size: cover;
    overflow: hidden;
    height: 190px;
  }

  @include element(name) {
    margin: 0 0 10px 0;
  }

  @include element(action-bar) {
    margin: 0 0 20px 0;
  }

  @include element(description) {
    margin: 0;
  }

  @include element(body) {
    background: $white;
    padding: 20px 30px;
    flex: 2;
  }
}
