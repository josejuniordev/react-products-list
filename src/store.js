import { applyMiddleware, compose, createStore } from "redux";
import rootReducer from "./rootReducer";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;


const store = createStore(
  rootReducer
);


export default store;
