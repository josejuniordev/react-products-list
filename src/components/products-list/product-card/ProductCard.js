import React from 'react';
import PropTypes from 'prop-types';
import Product from '../../../classes/Product';
import './ProductCard.scss';
import AmountAndFavoriteInfoBar from '../../amount-and-favorite-info-bar/AmountAndFavoriteInfoBar';
import ImageDisplay from '../../product-parts/image-display/ImageDisplay';

function ProductCard(
  {
    product,
    onClick = () => {},
  }
) {

  function onClickHandler(ev) {
    ev.preventDefault();
    onClick();
  }

  return (
    <article className='product-card'>
      <div className="product-card__image">
        <ImageDisplay product={product} />
      </div>
      <div className="product-card__body">
        <div className="product-card__action-bar">
          <AmountAndFavoriteInfoBar
            amount={product.amount}
            isFavorite={product.favorite}
            productId={product.id}
          />
        </div>
        <h2 className='product-card__name'><a href="#" onClick={onClickHandler}>{product.name}</a></h2>
        <p className='product-card__description'>{product.shortDescription}</p>
      </div>
    </article>
  )
}

ProductCard.propTypes = {
  product: PropTypes.instanceOf(Product),
  onClick: PropTypes.func,
};

export default ProductCard;
