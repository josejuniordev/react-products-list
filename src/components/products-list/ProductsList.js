import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import ProductCard from './product-card/ProductCard';
import Product from '../../classes/Product';
import './ProductsList.scss';

const amount = 3;

export function ProductsList(
  {
    products,
  },
) {
  const [loadedProducts, setLoadedProducts] = useState([]);
  const [firstUpdate, setFirstUpdate] = useState(true);
  const [keepTheState, setKeepTheState] = useState(false);
  const [loadMore, setLoadMore] = useState(false);
  const [limit, setLimit] = useState(amount);
  const [offset, setOffset] = useState(0);

  useEffect(() => {
    function loadProducts() {
      setLoadMore(false);
      let loadedProductsList;

      if (keepTheState) {
        setKeepTheState(false);
        const sliced = products.slice(0, limit);
        loadedProductsList = [...sliced];
      } else {
        if (loadedProducts.length >= products.length) {
          return;
        }

        const sliced = products.slice(offset, limit);
        loadedProductsList = [...loadedProducts, ...sliced];

        setOffset(offset + amount);
        setLimit(limit + amount);
      }

      setLoadedProducts(loadedProductsList);
    }

    if (loadMore) {
      loadProducts();
    }
  }, [loadMore, limit, loadedProducts, offset, products, keepTheState]);

  useEffect(() => {
    function handleProductsListUpdate() {
      if (firstUpdate) {
        setLoadMore(true);
        setFirstUpdate(false);
      } else {
        setKeepTheState(true);
        setLoadMore(true);
      }
    }

    if (products && products.length) {
      handleProductsListUpdate();
    } else {
      setLoadedProducts(products);
    }
  }, [products, firstUpdate]);

  useEffect(() => {
    const wrapper = document.querySelector('.layout-container');

    const timer = setInterval(() => {
      const scrollLimit = wrapper.offsetHeight - window.innerHeight;

      if (window.scrollY >= scrollLimit) {
        setLoadMore(true);
      }
    }, 500);

    return () => {
      window.clearInterval(timer);
    };
  }, []);

  return (
    <>
      <section className="products-list">
        {
          loadedProducts && loadedProducts.length
            ? (
              loadedProducts.map((product) => (
                <div key={product.id} className="products-list__item">
                  <ProductCard key={product.id} product={product} />
                </div>
              ))
            )
            : <p>NÃ£o existem produtos para exibir.</p>
        }
      </section>
      {
        loadMore && <p>carregando ...</p>
      }
    </>
  );
}

ProductsList.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.instanceOf(Product),
  ),
};

ProductsList.defaultProps = {
  products: [],
};

export default withRouter(ProductsList);
