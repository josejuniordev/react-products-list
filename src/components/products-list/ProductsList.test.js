import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { ProductsList } from './ProductsList';
import { productsMock } from '../../mock/products-mock';
import ProductCard from './product-card/ProductCard';
import Product from '../../classes/Product';
import store from '../../store';
import history from '../../history';

describe('<ProductsList />', () => {
  describe('with products', () => {
    const products = productsMock.map((product) => new Product(product));
    let testableMethods;

    const renderedWrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ProductsList giveTestAccess={(values) => testableMethods = values} products={products} />
        </Router>
      </Provider>,
    );

    it('should render correctly', () => {
      expect(renderedWrapper).toMatchSnapshot();
    });

    it('should render the correctly number of cards to number of products', () => {
      const infiniteLoaderInitialNumber = 3;

      expect(renderedWrapper.find(ProductCard).length).toBe(infiniteLoaderInitialNumber);
    });
  });

  describe('without products', () => {
    const renderedWrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ProductsList products={[]} />
        </Router>
      </Provider>,
    );

    it('should render correctly', () => {
      expect(renderedWrapper).toMatchSnapshot();
    });

    it('should render no products', () => {
      expect(renderedWrapper.find(ProductCard).length).toBe(0);
    });
  });
});
