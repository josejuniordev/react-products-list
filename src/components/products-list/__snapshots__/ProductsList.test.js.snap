// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsList /> with products should render correctly 1`] = `ReactWrapper {}`;

exports[`<ProductsList /> without products should render correctly 1`] = `ReactWrapper {}`;
